swift/apple/FirezoneNetworkExtension/SessionManager.swift

```diff
@@ -0,0 +1,250 @@
+        }
+
+        Log.log("SessionManager: Command handling finished")
+      }
+
+      // Wait for first task to complete, then cancel all
+      _ = await group.next()
+      Log.log("SessionManager: One task completed, cancelling event loop")
+      group.cancelAll()
+    }
+
+    // Cleanup when event loop exits
+    // Assigning to `nil` will invoke `Drop` on the Rust side
+    // Do NOT call disconnect() explicitly - let Drop handle everything
+    Log.log("SessionManager: Event loop finished, cleaning up session")
+    self.session = nil
+    Log.log("SessionManager: Session cleaned up")
+  }
+
+  /// Handles a command by calling the appropriate session method.
+  private func handleCommand(_ command: SessionCommand, session: Session) async {
+    switch command {
+    case .disconnect:
+      Log.log("SessionManager: Handling disconnect command")
+      // Assigning to `nil` will invoke `Drop` on the Rust side
+      // Do NOT call disconnect() explicitly - let Drop handle everything
+      self.session = nil
+
+    case .setDisabledResources(let resources):
+      Log.log("SessionManager: Handling setDisabledResources command")
+      do {
+        try session.setDisabledResources(disabledResources: resources)
+      } catch {
+        Log.error(error)
+      }
+
+    case .setDns(let servers):
+      Log.log("SessionManager: Handling setDns command")
+      do {
+        try session.setDns(dnsServers: servers)
+      } catch {
+        Log.error(error)
+      }
+
+    case .reset(let reason):
+      Log.log("SessionManager: Handling reset command: \(reason)")
+      session.reset(reason: reason)
+    }
+  }
+
+  /// Sends a command to the session via the command channel.
+  ///
+  /// This is safe to call from any thread/actor context.
+  nonisolated func sendCommand(_ command: SessionCommand) {
+    commandContinuation.yield(command)
+  }
+
+  /// Sets the TUN device by searching for it (async wrapper for setTunFromSearch).
+  func setTunFromSearch() async throws {
+    guard let session = session else {
+      throw AdapterError.invalidSession(nil)
+    }
+    try session.setTunFromSearch()
+  }
+
+  /// Clears the session, triggering Drop on the Rust side.
+  ///
+  /// This must happen asynchronously to allow Rust to break cyclic dependencies
+  /// between the runtime and the task that is executing the callback.
+  func clearSession() async {
+    // Assigning to `nil` will invoke `Drop` on the Rust side
+    Log.log("SessionManager: Clearing session")
+    self.session = nil
+  }
+
+  /// Stops the session by sending a disconnect command and cancelling the event loop.
+  func stop() async {
+    Log.log("SessionManager: Stopping")
+    sendCommand(.disconnect)
+
+    // Wait briefly for graceful shutdown
+    try? await Task.sleep(nanoseconds: 100_000_000)  // 100ms
+
+    // Cancel event loop task
+    eventLoopTask?.cancel()
+    eventLoopTask = nil
+
+    // Close command stream
+    commandContinuation.finish()
```

Comment:
Coming from Rust, I think this should be called `sender`.

Reply:
based on OOB discussion, I'll wrap in something that resembles a Rust channel.